File: 2019/day03/day03.hs
#!/usr/bin/env stack
-- stack --resolver lts-14.16 script
module Day03 where

import System.Environment (getArgs)
import qualified Data.Map as M
import Data.List (minimum)
import Data.Function (on)
import Text.Parsec
import Text.Parsec.Char
import Text.Parsec.String

data Dir = U | D | L | R

parseDir :: Parser Dir
parseDir =   (char 'U' >> return U)
         <|> (char 'D' >> return D)
         <|> (char 'L' >> return L)
         <|> (char 'R' >> return R)

parseSeg :: Parser (Dir, Int)
parseSeg = do
  d <- parseDir
  i <- read <$> many digit
  return (d, i)

parseWire :: Parser [(Dir, Int)]
parseWire = parseSeg `sepBy` char ','

delta :: Dir -> (Int, Int)
delta U = ( 0, -1)
delta D = ( 0,  1)
delta L = (-1,  0)
delta R = ( 1,  0)

wirePositions :: [(Dir, Int)] -> M.Map (Int, Int) Int
wirePositions = M.fromList . reverse . go 0 0 0
  where go _ _ _ [] = []
        go x y s ((_, 0):w) = go x y s w
        go x y s ((d, n):w) = ((x', y'), s') : go x' y' s' ((d, n - 1):w)
          where (dx, dy) = delta d
                x' = x + dx
                y' = y + dy
                s' = s + 1

star1 :: M.Map (Int, Int) Int -> M.Map (Int, Int) Int -> Int
star1 ma mb = minimum . map snd . M.toList $ dists
  where dists = M.intersectionWithKey (\(x, y) _ _ -> abs x + abs y) ma mb

star2 :: M.Map (Int, Int) Int -> M.Map (Int, Int) Int -> Int
star2 ma mb = minimum . map snd . M.toList $ steps
  where steps = M.intersectionWith (+) ma mb

main :: IO ()
main = do
  args <- getArgs
  let
    filename = case args of
      (x:_) -> x
      _     -> "input.txt"
  [Right a, Right b] <- map (parse parseWire "") . lines <$> readFile filename
  let ma = wirePositions a
  let mb = wirePositions b
  print $ star1 ma mb
  print $ star2 ma mb

File: 2019/day03/day03.rs
#!/usr/bin/env run-cargo-script
use std::collections::HashMap;
use std::collections::HashSet;

fn parse_wire(s: &str) -> Vec<((i32, i32), i32)> {
    s.split(",")
        .map(|i| {
            let (dir, len) = i.split_at(1);
            let delta = match dir {
                "U" => (0, -1),
                "D" => (0, 1),
                "L" => (-1, 0),
                "R" => (1, 0),
                _ => panic!(format!("Unknown direction {}", dir)),
            };
            (delta, len.parse().unwrap())
        })
        .collect()
}

fn get_wire_positions(wire: &Vec<((i32, i32), i32)>) -> HashMap<(i32, i32), i32> {
    let mut res = HashMap::new();
    let mut x = 0;
    let mut y = 0;
    let mut steps = 0;
    for ((dx, dy), length) in wire {
        for _ in 0..*length {
            x += dx;
            y += dy;
            steps += 1;
            res.entry((x, y)).or_insert(steps);
        }
    }
    res
}

fn get_wire_intersections(
    w1: &HashMap<(i32, i32), i32>,
    w2: &HashMap<(i32, i32), i32>,
) -> HashSet<(i32, i32)> {
    let s1: HashSet<_> = w1.keys().collect();
    s1.intersection(&w2.keys().collect()).map(|i| **i).collect()
}

fn star1(intersections: &HashSet<(i32, i32)>) -> i32 {
    intersections
        .iter()
        .map(|(x, y)| x.abs() + y.abs())
        .min()
        .unwrap()
}

fn star2(
    w1: &HashMap<(i32, i32), i32>,
    w2: &HashMap<(i32, i32), i32>,
    intersections: &HashSet<(i32, i32)>,
) -> i32 {
    intersections.iter().map(|i| w1[i] + w2[i]).min().unwrap()
}

fn main() {
    let args: Vec<_> = std::env::args().collect();
    let filename = args.get(1).map(|s| &s[..]).unwrap_or("input.txt");
    let file_contents = std::fs::read_to_string(filename).unwrap();
    let mut input = file_contents.split_whitespace();

    let w1 = get_wire_positions(&parse_wire(input.next().unwrap()));
    let w2 = get_wire_positions(&parse_wire(input.next().unwrap()));
    let intersections = get_wire_intersections(&w1, &w2);

    println!("{}", star1(&intersections));
    println!("{}", star2(&w1, &w2, &intersections));
}
